### ISSUE TITLE
IfcComplexProeprty

### ISSUE THREAD

**Issue by epeter-** [Autodesk/revit-ifc#267]
since the latest release of revit-ifc (21.2.0.0) there are several new features / improvements to ifc for revit. 
one is the implementation of:

**Improved handling of Complex Property in the PropertySet definition**

(how) is it possible to create and export complex properties?

**Comment by WawanSolihin** [comment]
This has not been fully tested, but I think currently it supports IFC PSet_* with complex property in it. The Revit parameter that will map to the complex property in IFC will take the following form: <Complex property name>.<Property name>
For example in IFC2x3: Pset_Draughting has Colour defined as a complex property containing: Red, Green and Blue. The parameter map in Revit for it will be: Colour.Red, Colour.Green, and Colour.Blue.
We do not have a generic way yet to define and handle the complex property currently. It also seems that in IFC4 the use of complex property has been reduced. If there is a great demand for it, we may need to look into a future implementation to support it.

**Comment by epeter-** [comment]
wawan, i tried to understand what you suggested but failed.
what i did was:
create a revit model
![grafik](https://user-images.githubusercontent.com/6928491/104764981-6f45e900-5768-11eb-88b5-96e529b1c0de.png)
as you see there are 3 properties 
Farbe.Rot
Farbe.Grün
Farbe.Blau
(meaning colour.red, blue and green in german language)

i created a export definition file for testing with various approaches:

PropertySet:	PSet_Complex1	I	IfcWall
	Colour	Integer	Farbe.Rot
	Colour	Integer	Farbe.Grün
	Colour	Integer	Farbe.Blau
PropertySet:	PSet_Complex2	I	IfcWall
	Farbe	Integer	Farbe.Rot
	Farbe	Integer	Farbe.Grün
	Farbe	Integer	Farbe.Blau
PropertySet:	PSet_Complex3	I	IfcWall
	Farbe.Rot	Integer	
	Farbe.Grün	Integer	
	Farbe.Blau	Integer	

PropertySet:	PSet_Complex4	I	IfcWall
	Farbe	Text

PropertySet:	PSet_Complex5	I	IfcWall
	Colour.Red	Integer	Farbe.Rot
	Colour.Green	Integer	Farbe.Grün
	Colour.Blue	Integer	Farbe.Blau

PropertySet:	PSet_Complex6	I	IfcWall
	Colour	Text

unfortuntely i do not receive the result expected only:

![grafik](https://user-images.githubusercontent.com/6928491/104765277-ed09f480-5768-11eb-9cf2-f91baf7b9f84.png)

what am i missing?
tx peter

**Comment by WawanSolihin** [comment]
The limitation currently is that it only supports IFC built-in Psets with complex property. There is no support yet for user-defined propertysets due to complexity to capture the definition.

**Comment by epeter-** [comment]
@WawanSolihin  @AngelVelezSosa 
wawan, i  want to suggest a possible solution for defining user-defined complex properties:
![image](https://user-images.githubusercontent.com/6928491/121338259-a090a280-c91d-11eb-8b7d-5666423322d9.png)

this could be actually pretty easy as you can see in attached picture.
only prerequisite is that a complex property has to be defined before being used in a pset.
what do you think?
peter

**Comment by WawanSolihin** [comment]
Thanks Peter for the suggestion.
We will register this as an improvement in our backlog.

**Comment by epeter-** [comment]
autodesk is well known for prioritizing customer requests at its own discretion. as a result it can happen that a topic remains in "open" status for many, many years.

**Comment by AngelVelezSosa** [comment]
Unfortunately, we don't have the resources to address all possible improvements immediately. We prioritize bug fixes first.  For things like parameter mapping, our priority is to improve the user experience so that we can get away from using text files to control export functionality.   If you would like a more interactive way of influencing our roadmap, I recommend you join the Preview Release program where you can see what we are actively working on.

**Comment by epeter-** [comment]
angel you have to admit, however, that even taking into account the priorities, 4 years is too long to process customers suggestion/request.
in my opinion, it would be fairer to make it clear that there are no resources available for this improvement instead of adding to a backlog.

**Comment by AngelVelezSosa** [comment]
You are correct that we didn't have the right resources to manage this GitHub in addition to all of the other sources of requests- direct email, other social media, official support issues, Revit ideas and of course projects.  We are now systematically going through all open issues and at least getting them into our internal system if they aren't there already and I hope that once we catch up we can upgrade this to a first class issue reporting source. 

So far, I haven't closed anything as effectively "won't fix", but once we've cataloged everything, if I think it isn't on the 3 year map, I'll explicitly close it with an explanation to give people a chance to try to escalate if they think I've got it wrong.  That's the plan, anyway.
